### DISCUSSION TITLE
Use a more consistent sun and moon logic regardless of source

### DISCUSSION THREAD

**Discussion by elydpg** [breezy-weather/breezy-weather#1115] [General]
Spiritual successor to #1106. In that discussion, I proposed a more consistent logic for the sunrise and sunset times:

> here are all of the logical possibilities for a given calendar day:
> 
> 1. There are no sunsets or sunrises. In this case, it needs to be determined another way whether it is day or night in order to apply the theme accordingly. It might not make sense to render the sun in this case, even during daytime.
> 2. There is exactly one sunrise. In this case the time before sunrise is night time and no sun should be rendered. After sunrise, render the sun in position based on the time until the earliest sunset in the next calendar day.
> 3. There is exactly one sunset. In this case the time after sunset is night time, so no sun render, and before sunset render the sun in position based on the time since the latest sunrise in the previous calendar day.
> 4. There is sunrise followed by sunset. This is the typical case; after sunset and before sunrise is night time, in between is daytime. Render the sun accordingly.
> 5. There is sunset followed by sunrise. In this case the time between sunset and sunrise is night time so no render. The time before sunset is daytime so do as in case 3, and the time after sunrise is also daytime so do as in case 2
> 6. There is sunset followed by sunrise followed by sunset. In this case do as in case 3 before the first sunset, otherwise do as in case 4.
> 7. There is sunrise followed by sunset followed by sunrise. In this case do as in case 2 after the last sunrise, otherwise do as in case 4.
> 
> The same logic applies to the moonrise and moonset, with the exception that there is no need to determine the theme, only the moon render position. I'm fairly sure as well that cases 6 and 7 don't happen for the moon, as the period between moonrises and moonsets is always about 25 hours.

I was originally going to suggest that the sun and moon logic could be made more consistent with the help of a 3rd party library instead of using the source data: [commons-suncalc](https://github.com/shred/commons-suncalc) was my suggestion, but looking at the code I see that is how the fallback computation is done anyway, so I'm not sure how viable it would be to use only commons-suncalc for the sun and moon logic. I also realize that my original logic is flawed slightly (case 2 does not work if the next day has no sunrises or sunsets).

Anyway I'm leaving this open for further discussion on the matter :)

**Comment by papjul** [comment]
> I'm not sure how viable it would be to use only commons-suncalc for the sun and moon logic

It should not be used in priority:

> Astronomical calculations are far more complex than throwing a few numbers into an obscure formula and then getting a fully accurate result. There is always a tradeoff between accuracy and computing time.
>
> This library has its focus on getting acceptable results at low cost, so it can also run on mobile devices, or devices with a low computing power. The results have an accuracy of about a minute, which should be good enough for common applications (like sunrise/sunset timers), but is probably not sufficient for astronomical purposes.
>
> If you are looking for the highest possible accuracy, you are looking for a different library.

I noticed slight differences with moon phases on this library than other multiple sources. And I trust sources to have equal or better data.
